modules/pipeline/providers/reconciler/task_reconciler.go

```diff
@@ -0,0 +1,391 @@
+// Copyright (c) 2021 Terminus, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reconciler
+
+import (
+	"context"
+	"time"
+
+	"github.com/xormplus/xorm"
+
+	"github.com/erda-project/erda-infra/base/logs"
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/modules/pipeline/aop"
+	"github.com/erda-project/erda/modules/pipeline/aop/aoptypes"
+	"github.com/erda-project/erda/modules/pipeline/dbclient"
+	"github.com/erda-project/erda/modules/pipeline/pipengine/actionexecutor"
+	"github.com/erda-project/erda/modules/pipeline/pipengine/actionexecutor/types"
+	"github.com/erda-project/erda/modules/pipeline/pkg/errorsx"
+	"github.com/erda-project/erda/modules/pipeline/providers/reconciler/rutil"
+	"github.com/erda-project/erda/modules/pipeline/providers/reconciler/taskpolicy"
+	"github.com/erda-project/erda/modules/pipeline/providers/reconciler/taskrun"
+	taskop2 "github.com/erda-project/erda/modules/pipeline/providers/reconciler/taskrun/taskop"
+	"github.com/erda-project/erda/modules/pipeline/spec"
+	"github.com/erda-project/erda/pkg/loop"
+	"github.com/erda-project/erda/pkg/strutil"
+)
+
+type TaskReconciler interface {
+	// ReconcileOneTaskUntilDone reconcile one task until done.
+	// done means end status, include success, failed and others.
+	ReconcileOneTaskUntilDone(ctx context.Context, p *spec.Pipeline, task *spec.PipelineTask)
+
+	SaveTaskIdempotently(ctx context.Context, p *spec.Pipeline, task *spec.PipelineTask) error
```

Comment:
is `IdempotentSaveTask` more proper? @shuofan

Reply:
adjusted to `IdempotentSaveTask`